File: apps/android/app/src/main/java/com/exponential/app/ui/issue/DiffRendering.kt
package com.exponential.app.ui.issue

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.ScrollState
import androidx.compose.foundation.background
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.selection.SelectionContainer
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawWithContent
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Rect
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.BlendMode
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Paint
import androidx.compose.ui.graphics.drawscope.clipRect
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.exponential.app.domain.Diff
import com.exponential.app.ui.theme.DesignTokens
import com.exponential.app.ui.theme.Motion
import com.exponential.app.ui.theme.TextEmphasis

// EXP-895 — the patch BODY of the one diff view ×4 (web
// `packages/ui/src/file-diff-card.tsx`, desktop `crates/ui/src/diff`, iOS
// `DiffRendering.swift`). It renders a parsed [Diff.Hunk] list and nothing
// else: no raw patch strings, no per-line sign sniffing — the ONE parser
// (`domain/Diff.kt`) already decided what every line is.
//
// Unified layout only, four columns: old gutter · new gutter · sign · text.
// `compact` (a transcript card) drops the OLD gutter — the column a 320dp-wide
// tool row can least afford — and keeps everything else identical.
//
// Every colour comes from `DesignTokens.Diff.*` (generated from
// `packages/design-tokens/tokens.json`), never from a hand-picked emerald.

/**
 * What a file with no hunks says instead of rows — a binary blob, a pure
 * rename, an empty new file and a patch GitHub refused to send all land here,
 * and the reader has to be told WHICH. Byte-identical to web `noHunksNote`.
 */
fun noHunksNote(file: Diff.File): String = when {
    file.binary -> "Binary file"
    file.status == Diff.Status.ADDED -> "Empty file added"
    file.status == Diff.Status.REMOVED -> "File removed"
    file.status == Diff.Status.RENAMED -> "Renamed without content changes"
    file.status == Diff.Status.COPIED -> "Copied without content changes"
    else -> "No textual diff (binary or too large)"
}

/**
 * Rendered-line cap per file — every line composes a Text under
 * IntrinsicSize.Max intrinsic measurement, so an uncapped multi-thousand-line
 * patch (lockfile PRs, raw worktree diffs) freezes the frame. The web expands
 * in `LINE_CHUNK` steps instead; a phone card simply stops and says so.
 */
const val DIFF_MAX_RENDERED_LINES = 600

/**
 * Rows per fade band. The trailing fade is an alpha MASK, so the masked pixels
 * have to live in an offscreen layer — and a layer spanning a whole 600-line
 * patch is viewport-wide by ~23k px tall, past the GPU max texture size, where
 * HWUI silently drops the layer and the DstIn rect punches a hole through the
 * window instead. Banding the rows keeps every layer a few hundred px tall.
 */
private const val DIFF_FADE_BAND_ROWS = 40

/** One display row of a file's patch (web `Row`, same three kinds). */
sealed interface DiffRow {
    /** A `N unchanged lines` divider — context the patch never carried. */
    data class Gap(val text: String) : DiffRow

    /** The verbatim `@@ … @@` header, section heading and all. */
    data class Header(val text: String) : DiffRow

    data class Body(val line: Diff.Line) : DiffRow
}

/** [buildDiffRows]' answer: the rows to draw, and the lines that did not fit. */
data class DiffRows(val rows: List<DiffRow>, val hidden: Int)

/**
 * A file's hunks flattened into display rows: a gap divider before each hunk
 * that skipped context (including the file's own head, `unchangedBefore`), the
 * `@@` header, then the hunk's lines — stopping dead at [maxLines] BODY rows,
 * exactly where the web's reveal cap stops.
 *
 * Pure, so `DiffRowsTest` can hold the shape without a device.
 */
fun buildDiffRows(hunks: List<Diff.Hunk>, maxLines: Int = DIFF_MAX_RENDERED_LINES): DiffRows {
    var total = 0
    for (hunk in hunks) total += hunk.lines.size
    val out = mutableListOf<DiffRow>()
    var shown = 0
    outer@ for ((index, hunk) in hunks.withIndex()) {
        val skipped = if (index == 0) {
            Diff.unchangedBefore(hunk)
        } else {
            Diff.unchangedBetween(hunks[index - 1], hunk)
        }
        // A PLAIN row, never a button: the skipped context is not on the wire,
        // so there is nothing to expand to (EXP-895).
        if (skipped > 0) out.add(DiffRow.Gap(Diff.unchangedLabel(skipped)))
        out.add(DiffRow.Header(hunk.header))
        for (line in hunk.lines) {
            if (shown >= maxLines) break@outer
            out.add(DiffRow.Body(line))
            shown += 1
        }
    }
    return DiffRows(out, maxOf(0, total - shown))
}

private val DiffFontSize = 11.sp
private val DiffLineHeight = 15.sp

/** The two number columns; the compact card keeps only the new one. */
private val GutterWidth: Dp = 30.dp
private val CompactGutterWidth: Dp = 26.dp

/** The sign column — one glyph wide, `+` / U+2212 / nothing. */
private val SignWidth: Dp = 12.dp

/**
 * The monospace patch body of ONE file: `@@` headers on the hunk band,
 * `N unchanged lines` dividers between hunks, and the lines themselves on
 * their add/del washes. Horizontal scrolling lives INSIDE this block — never
 * on the page — and the trailing edge fades while there is more to the right.
 */
@Composable
fun PatchLines(
    hunks: List<Diff.Hunk>,
    modifier: Modifier = Modifier,
    /** A transcript card: drops the old-side gutter, keeps the rest. */
    compact: Boolean = false,
    maxLines: Int = DIFF_MAX_RENDERED_LINES,
) {
    val built = remember(hunks, maxLines) { buildDiffRows(hunks, maxLines) }
    val scrollState = rememberScrollState()
    val fadeStrength = trailingFadeStrength(scrollState)
    val contextColor = MaterialTheme.colorScheme.onSurface.copy(alpha = TextEmphasis.Secondary)
    val gutter = if (compact) CompactGutterWidth else GutterWidth
    Column(modifier = modifier.fillMaxWidth()) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .horizontalScroll(scrollState),
        ) {
            SelectionContainer {
                Column(modifier = Modifier.width(IntrinsicSize.Max)) {
                    built.rows.chunked(DIFF_FADE_BAND_ROWS).forEach { band ->
                        Column(
                            modifier = Modifier
                                .fillMaxWidth()
                                .trailingScrollFade(scrollState, fadeStrength),
                        ) {
                            band.forEach { row ->
                                PatchRow(
                                    row = row,
                                    compact = compact,
                                    gutter = gutter,
                                    contextColor = contextColor,
                                )
                            }
                        }
                    }
                }
            }
        }
        if (built.hidden > 0) {
            Text(
                text = "Diff truncated. Showing the first $maxLines lines.",
                color = contextColor,
                fontSize = DiffFontSize,
                lineHeight = DiffLineHeight,
                modifier = Modifier.padding(horizontal = 10.dp, vertical = 6.dp),
            )
        }
    }
}

@Composable
private fun PatchRow(row: DiffRow, compact: Boolean, gutter: Dp, contextColor: Color) {
    when (row) {
        // The divider sits on a WEAKER wash than the `@@` header above it — it
        // is the absence of a hunk, not one.
        is DiffRow.Gap -> Text(
            text = row.text,
            color = DesignTokens.Diff.GutterFg,
            fontFamily = FontFamily.Monospace,
            fontSize = DiffFontSize,
            lineHeight = DiffLineHeight,
            maxLines = 1,
            softWrap = false,
            textAlign = TextAlign.Center,
            modifier = Modifier
                .fillMaxWidth()
                .background(GapFill)
                .padding(horizontal = 10.dp),
        )
        is DiffRow.Header -> Text(
            text = row.text,
            color = DesignTokens.Diff.HunkFg,
            fontFamily = FontFamily.Monospace,
            fontSize = DiffFontSize,
            lineHeight = DiffLineHeight,
            maxLines = 1,
            softWrap = false,
            modifier = Modifier
                .fillMaxWidth()
                .background(DesignTokens.Diff.HunkBg)
                .padding(horizontal = 10.dp),
        )
        is DiffRow.Body -> PatchBodyRow(row.line, compact, gutter, contextColor)
    }
}

/** `bg-diff-hunk-bg/60` — the gap divider's share of the hunk band. */
private val GapFill: Color =
    DesignTokens.Diff.HunkBg.copy(alpha = DesignTokens.Diff.HunkBg.alpha * 0.6f)

@Composable
private fun PatchBodyRow(line: Diff.Line, compact: Boolean, gutter: Dp, contextColor: Color) {
    // `\ No newline at end of file` — numbered on NEITHER side, so it gets no
    // gutters at all and leans italic to read as metadata rather than content.
    if (line.kind == Diff.LineKind.META) {
        Text(
            text = line.text.ifEmpty { " " },
            color = DesignTokens.Diff.GutterFg,
            fontFamily = FontFamily.Monospace,
            fontStyle = FontStyle.Italic,
            fontSize = DiffFontSize,
            lineHeight = DiffLineHeight,
            maxLines = 1,
            softWrap = false,
            modifier = Modifier.fillMaxWidth().padding(horizontal = 10.dp),
        )
        return
    }
    val fill = when (line.kind) {
        Diff.LineKind.ADD -> DesignTokens.Diff.AddBg
        Diff.LineKind.DEL -> DesignTokens.Diff.DelBg
        else -> Color.Transparent
    }
    val ink = when (line.kind) {
        Diff.LineKind.ADD -> DesignTokens.Diff.AddFg
        Diff.LineKind.DEL -> DesignTokens.Diff.DelFg
        else -> contextColor
    }
    Row(modifier = Modifier.fillMaxWidth().background(fill)) {
        if (!compact) Gutter(line.oldNo, gutter)
        Gutter(line.newNo, gutter)
        Text(
            // U+2212 MINUS SIGN, never a hyphen — the same glyph the counts
            // use, locked ×4.
            text = when (line.kind) {
                Diff.LineKind.ADD -> "+"
                Diff.LineKind.DEL -> "−"
                else -> " "
            },
            color = ink,
            fontFamily = FontFamily.Monospace,
            fontSize = DiffFontSize,
            lineHeight = DiffLineHeight,
            maxLines = 1,
            softWrap = false,
            textAlign = TextAlign.Center,
            modifier = Modifier.width(SignWidth),
        )
        Text(
            text = line.text.ifEmpty { " " },
            color = ink,
            fontFamily = FontFamily.Monospace,
            fontSize = DiffFontSize,
            lineHeight = DiffLineHeight,
            maxLines = 1,
            softWrap = false,
            modifier = Modifier.padding(end = 10.dp),
        )
    }
}

/** One line-number column: right-aligned, dimmed, never selectable noise. */
@Composable
private fun Gutter(number: Int?, width: Dp) {
    Text(
        text = number?.toString().orEmpty(),
        color = DesignTokens.Diff.GutterFg,
        fontFamily = FontFamily.Monospace,
        fontSize = DiffFontSize,
        lineHeight = DiffLineHeight,
        maxLines = 1,
        softWrap = false,
        textAlign = TextAlign.End,
        modifier = Modifier.width(width).padding(end = 4.dp),
    )
}

/**
 * Wide enough to dissolve a few characters, narrow enough to leave the line
 * readable — the horizontal twin of the steer feed's top fade (iOS
 * `DiffPatchBlock.trailingFadeWidth` parity).
 */
private val TrailingFadeWidth: Dp = 36.dp

/**
 * EXP-722: a scroller with no scrollbar has NO affordance at rest — a line cut
 * flush at the block's edge reads as truncated, not as "more to the right". So
 * the trailing edge FADES while [state] can still scroll forward and turns
 * crisp once the reader has panned to the end. Reduce Motion snaps the edge
 * instead of dissolving it. Hoisted out of [trailingScrollFade] so one patch
 * runs ONE animation, not one per band.
 */
@Composable
private fun trailingFadeStrength(state: ScrollState): Float {
    val strength by animateFloatAsState(
        targetValue = if (state.canScrollForward) 1f else 0f,
        animationSpec = Motion.fast(),
        label = "diff-trailing-fade",
    )
    return strength
}

/**
 * Paints the trailing fade over one band of scrolled patch rows.
 *
 * It is an alpha MASK, not a painted gradient: the block sits on translucent
 * glass over the page gradient, so no single colour would match what is behind
 * it. A DstIn rect only masks the patch when the pixels it multiplies live in
 * an offscreen layer, so the fade strip — and ONLY the strip — is wrapped in a
 * `saveLayer`: fade-width by band-height, small enough to always be a real
 * layer, versus the viewport-by-whole-patch layer that a `graphicsLayer` on
 * the scroller would demand (and that HWUI drops past the max texture size,
 * leaving DstIn to cut a hole through the window behind the patch). Nothing is
 * layered at all while [strength] is 0, i.e. while the band does not overflow.
 *
 * This modifier sits INSIDE [horizontalScroll], so the strip is placed in
 * content coordinates: the viewport's trailing edge is `value + viewportSize`.
 */
private fun Modifier.trailingScrollFade(state: ScrollState, strength: Float): Modifier =
    this.drawWithContent {
        val content = this
        val fadeWidth = TrailingFadeWidth.toPx()
        val right = (state.value + state.viewportSize).toFloat()
        val left = right - fadeWidth
        if (strength <= 0f || state.viewportSize == 0 || left >= size.width || left < 0f) {
            content.drawContent()
            return@drawWithContent
        }
        // Everything left of the strip draws straight to the canvas, unlayered.
        clipRect(right = left) { content.drawContent() }
        val strip = Rect(left, 0f, right.coerceAtMost(size.width), size.height)
        drawContext.canvas.saveLayer(strip, Paint())
        clipRect(left = left) { content.drawContent() }
        drawRect(
            brush = Brush.horizontalGradient(
                colors = listOf(Color.Black, Color.Black.copy(alpha = 1f - strength)),
                startX = left,
                endX = right,
            ),
            topLeft = Offset(left, 0f),
            size = Size(fadeWidth, size.height),
            blendMode = BlendMode.DstIn,
        )
        drawContext.canvas.restore()
    }

File: apps/ios/Exponential/UI/Issue/DiffRendering.swift
import ExpCore
import ExpUI
import SwiftUI

/// EXP-895 — the ONE unified-diff BODY, ×4. Everything about what a diff IS
/// (parsing, counts, the skipped-context arithmetic, every label) lives in
/// ExpCore `Diff` / `DiffPresentation`; this file is presentation only:
/// a `Diff.File`'s hunks as rows of `old · new · sign · text`, painted from
/// `DesignTokens.Diff.*` (never a raw `.green`/`.red`), with the `@@` header on
/// its own band and a plain `N unchanged lines` divider wherever the patch
/// skipped context.
///
/// Unified layout only — a split view on a phone is two unreadable columns.
/// `compact` (a transcript's tool card) drops the OLD gutter and tightens the
/// type; it never changes WHAT is drawn.
///
/// The dividers are plain rows, never buttons: the skipped context is not on
/// the wire, so there is nothing to expand to.
struct DiffPatchBlock: View {
    let file: Diff.File
    var compact = false

    /// Wide enough to dissolve a few characters, narrow enough to leave the
    /// line readable — the horizontal twin of `StickyHeaderFade.height`.
    static let trailingFadeWidth: CGFloat = 36

    /// The layout cap. A 20k-line file must not be handed to SwiftUI as 20k
    /// rows; past this the block says so and stops. (A publisher's OWN cut is a
    /// different fact, reported by `Diff.Parsed.truncatedLines` above this.)
    static let maxRows = 600

    @Environment(\.motion) private var motion
    @State private var viewportWidth: CGFloat = 0
    /// The content's trailing edge in the scroller's coordinate space — it
    /// shrinks as the reader pans right. Seeded past any real width so the
    /// FIRST frame already fades: measured at 0 the block would render crisp
    /// and then flash into the fade one geometry pass later.
    @State private var contentTrailingEdge: CGFloat = .greatestFiniteMagnitude

    private var overflowsTrailing: Bool {
        contentTrailingEdge > viewportWidth + 1
    }

    // MARK: - Rows

    /// One display row. A `gap` is the divider, a `hunk` the verbatim `@@` line,
    /// a `line` one of the file's own rows.
    private enum Row: Identifiable {
        case gap(id: String, text: String)
        case hunk(id: String, text: String)
        case line(id: String, Diff.Line)

        var id: String {
            switch self {
            case let .gap(id, _): id
            case let .hunk(id, _): id
            case let .line(id, _): id
            }
        }
    }

    /// The file flattened: a divider before each hunk that skipped context (the
    /// FIRST hunk measures against the top of the file), the header, the lines.
    private static func rows(of file: Diff.File) -> [Row] {
        var out: [Row] = []
        for (index, hunk) in file.hunks.enumerated() {
            let skipped = index == 0
                ? Diff.unchangedBefore(hunk)
                : Diff.unchangedBetween(file.hunks[index - 1], hunk)
            if skipped > 0 {
                out.append(.gap(id: "g\(index)", text: Diff.unchangedLabel(skipped)))
            }
            out.append(.hunk(id: "h\(index)", text: hunk.header))
            for (at, line) in hunk.lines.enumerated() {
                out.append(.line(id: "l\(index)-\(at)", line))
            }
        }
        return out
    }

    // MARK: - Metrics
    //
    // The font is measured, not guessed: the row washes have to span the whole
    // block (a wash that stops at the end of the text reads as a second cut),
    // and a monospaced advance is the only way to give every row the same width
    // without padding the text with spaces.

    private struct Metrics {
        let font: Font
        let advance: CGFloat
        let gutter: CGFloat
        let sign: CGFloat
        let rowHeight: CGFloat
    }

    private var metrics: Metrics {
        // Dynamic type still drives the size — a transcript card just sits one
        // point below the page's own diff.
        let size = UIFont.preferredFont(forTextStyle: .caption2).pointSize - (compact ? 1 : 0)
        let uiFont = UIFont.monospacedSystemFont(ofSize: size, weight: .regular)
        let advance = ("0" as NSString).size(withAttributes: [.font: uiFont]).width
        return Metrics(
            font: Font(uiFont),
            advance: advance,
            // Four digits plus breathing room — the line numbers are
            // right-aligned inside it (web's `text-right tabular-nums`).
            gutter: advance * 4 + 8,
            sign: advance * 1.5,
            rowHeight: uiFont.lineHeight + 2
        )
    }

    /// The widest row in characters — the block's content width.
    private func contentColumns(_ rows: [Row]) -> Int {
        var widest = 0
        for row in rows {
            let count: Int
            switch row {
            case let .gap(_, text): count = text.count
            case let .hunk(_, text): count = text.count
            case let .line(_, line): count = line.text.count
            }
            if count > widest { widest = count }
        }
        return widest
    }

    var body: some View {
        let all = Self.rows(of: file)
        let shown = Array(all.prefix(Self.maxRows))
        let m = metrics
        let gutters = (compact ? m.gutter : m.gutter * 2) + m.sign
        let textWidth = max(CGFloat(contentColumns(shown)) * m.advance, 1)
        let rowWidth = max(gutters + textWidth, viewportWidth)

        VStack(alignment: .leading, spacing: 0) {
            if file.hunks.isEmpty {
                Text(DiffPresentation.noHunksNote(file))
                    .font(.caption)
                    .foregroundStyle(.white.opacity(TextOpacity.secondary))
                    .padding(.horizontal, 12)
                    .padding(.vertical, 8)
            } else {
                ScrollView(.horizontal, showsIndicators: false) {
                    VStack(alignment: .leading, spacing: 0) {
                        ForEach(shown) { row in
                            rowView(row, m: m, width: rowWidth, gutters: gutters)
                        }
                    }
                    .textSelection(.enabled)
                    .onGeometryChange(for: CGFloat.self, of: { $0.frame(in: .scrollView).maxX }) { edge in
                        contentTrailingEdge = edge
                    }
                }
                .onGeometryChange(for: CGFloat.self, of: { $0.size.width }) { width in
                    viewportWidth = width
                }
                .mask(trailingFadeMask)
                if all.count > shown.count {
                    Text("Diff truncated. Showing the first \(shown.count) lines.")
                        .font(.caption2)
                        .foregroundStyle(.white.opacity(TextOpacity.tertiary))
                        .padding(.horizontal, 8)
                        .padding(.vertical, 6)
                }
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
    }

    @ViewBuilder
    private func rowView(_ row: Row, m: Metrics, width: CGFloat, gutters: CGFloat) -> some View {
        switch row {
        case let .gap(_, text):
            // A PLAIN row, never a button (EXP-895).
            Text(text)
                .font(m.font)
                .foregroundStyle(DesignTokens.Diff.gutterFg)
                .frame(width: width, alignment: .center)
                .background(DesignTokens.Diff.hunkBg.opacity(0.6))
                .accessibilityIdentifier("diff-gap-row")
        case let .hunk(_, text):
            Text(text)
                .font(m.font)
                .foregroundStyle(DesignTokens.Diff.hunkFg)
                .lineLimit(1)
                .padding(.horizontal, 8)
                .frame(width: width, alignment: .leading)
                .background(DesignTokens.Diff.hunkBg)
        case let .line(_, line):
            if line.kind == .meta {
                // `\ No newline at end of file` — numbered on neither side.
                Text(line.text)
                    .font(m.font.italic())
                    .foregroundStyle(DesignTokens.Diff.gutterFg)
                    .lineLimit(1)
                    .padding(.horizontal, 8)
                    .frame(width: width, alignment: .leading)
            } else {
                HStack(spacing: 0) {
                    if !compact {
                        gutterCell(line.oldNo, m: m)
                    }
                    gutterCell(line.newNo, m: m)
                    Text(sign(line.kind))
                        .font(m.font)
                        .foregroundStyle(foreground(line.kind))
                        .frame(width: m.sign, alignment: .center)
                    Text(line.text.isEmpty ? " " : line.text)
                        .font(m.font)
                        .foregroundStyle(foreground(line.kind))
                        .lineLimit(1)
                        .fixedSize(horizontal: true, vertical: false)
                        .frame(width: max(width - gutters, 1), alignment: .leading)
                }
                .frame(width: width, alignment: .leading)
                .background(background(line.kind))
            }
        }
    }

    private func gutterCell(_ number: Int?, m: Metrics) -> some View {
        Text(number.map(String.init) ?? "")
            .font(m.font)
            .monospacedDigit()
            .foregroundStyle(DesignTokens.Diff.gutterFg)
            .lineLimit(1)
            .padding(.trailing, 4)
            .frame(width: m.gutter, alignment: .trailing)
    }

    private func sign(_ kind: Diff.LineKind) -> String {
        switch kind {
        case .add: "+"
        // U+2212 MINUS SIGN, the ×4 rule (`Diff.deletionsLabel`).
        case .del: "\u{2212}"
        default: " "
        }
    }

    private func foreground(_ kind: Diff.LineKind) -> Color {
        switch kind {
        case .add: DesignTokens.Diff.addFg
        case .del: DesignTokens.Diff.delFg
        case .context, .meta: .white.opacity(TextOpacity.secondary)
        }
    }

    private func background(_ kind: Diff.LineKind) -> Color {
        switch kind {
        case .add: DesignTokens.Diff.addBg
        case .del: DesignTokens.Diff.delBg
        case .context, .meta: .clear
        }
    }

    /// Opaque everywhere but the trailing strip, which fades to clear only
    /// while there is more to scroll to. Routed through the motion environment
    /// so Reduce Motion snaps the edge instead of dissolving it.
    private var trailingFadeMask: some View {
        HStack(spacing: 0) {
            Rectangle().fill(.black)
            LinearGradient(
                colors: [.black, .black.opacity(overflowsTrailing ? 0 : 1)],
                startPoint: .leading,
                endPoint: .trailing
            )
            .frame(width: Self.trailingFadeWidth)
        }
        .animation(motion.fast, value: overflowsTrailing)
    }
}
